### PR TITLE
Refactor map notes editing

### PR DESCRIPTION
**Pull request by AMurkin**
#### Summary
```SUMMARY: Features "Refactor map notes editing"```

#### Purpose of change
Currently, the user cannot use the left / right arrow keys to move the text cursor while editing the text of a note. This restores that ability with using separate windows to display a preview.

#### Describe the solution
Create input popup only once and show a preview in separate windows.
Added a simple map preview of the map tiles around the cursor (note position).
Used the input history feature.

Now:
default
PR. Showing both the note editing window and the preview area, and the popped-up history menu:
default

#### Describe alternatives you've considered
Not change a look, but fix text editing only.

### CODE REVIEW

**Comment by nsklaus** [comment]
no-one is saying anything about it, but i think it's nice added functionality.
i especially like note history and ability to use arrows in the note creation itself.

one idea: in your example screenshot, is the pink colored capital "L" in the middle of the screen a note ? if yes why i don't see it in the history ? if no why is the view not showing current note ?
is it covered by the opened windows ? it would be nice to either center the view  to make current note visible on the middle of the screen, or to be able to make the windows switch side so we can view what's under them, like the current note and its surroundings.

**Comment by AMurkin** [comment]
Thank you.
The ability to use arrows actually was here before, I just restored the normal behavior by replacing the preview display method.
Pink "L" - is a naturally spawned Lab.
History contains only 3 items because I added these entries after turning on the history. And you can erase the whole history at any time.
Yes, current note position on the map covered by the note editing window. I was thinking about shifting it to the side, but decided to do nothing about it at the moment.

**Comment by Uesrnme** [comment]
Thank you, cursor movement for note editing was very useful feature indeed.

Note string length limit is way too short now. 

And bracketing of the map cursor overlaps nearby map features, not helping to find current player location. Corner bracketing looks like roads. To see actual nearby terrain cursor have to be moved away.

It was better the old way.

**Comment by AMurkin** [comment]
I didn't change the length of the note string. You can always create notes near the first one.
To reduce confusion with the roads I set the borders color to yellow.
And I cannot move the input popup window from the center without major changes, I think.

File: src/overmap_ui.cpp [hunk 1]
```diff
@@ -859,59 +859,101 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
 
             const std::string old_note = overmap_buffer.note( curs );
             std::string new_note = old_note, tmp_note;
+            const int max_note_length = 45;
+            const bool has_debug_vision = g->u.has_trait( trait_id( "DEBUG_NIGHTVISION" ) );
+
+            std::vector<std::pair<nc_color, long>> map_around;
+            map_around.reserve( 9 );
```
Comment (codemime) [inline src/overmap_ui.cpp]: Fixed size implies an array, not a `vector`.

File: src/overmap_ui.cpp [hunk 2]
```diff
@@ -859,59 +859,101 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
...
+            for( int i = 1; i < 4; i++ ) {
+                for( int j = 1; j < 4; j++ ) {
+                    const int omy = curs.y + i - 2;
+                    const int omx = curs.x + j - 2;
```
Comment (codemime) [inline src/overmap_ui.cpp]: There's no need for manual traversal: `map::points_in_radius` will do this for you.

File: src/overmap_ui.cpp [hunk 3]
```diff
@@ -859,59 +859,101 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
...
+                    const bool see = has_debug_vision || overmap_buffer.seen( omx, omy, curs.z );
+
+                    nc_color ter_color = c_black;
+                    long ter_sym = ' ';
+                    if( see ) {
+                        // Only load terrain if we can actually see it
+                        oter_t const *info = nullptr;
+                        oter_id cur_ter = overmap_buffer.ter( omx, omy, curs.z );
+                        info = &cur_ter.obj();
+                        ter_color = info->get_color();
```
Comment (codemime) [inline src/overmap_ui.cpp]: `info` is a redundant temp: simple `cur_ter->get_color()` would work.

File: src/overmap_ui.cpp [hunk 4]
```diff
@@ -859,59 +859,101 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
...
+                        ter_sym = info->get_sym();
+                    } else {
+                        ter_color = c_dark_gray;
+                        ter_sym   = '#';
+                    }
+                    map_around.push_back( std::make_pair( ter_color, ter_sym ) );
```
Comment (codemime) [inline src/overmap_ui.cpp]: This can be boiled down to `map_around.emplace_back(ter_color, ter_sym)`.

File: src/overmap_ui.cpp [hunk 5]
```diff
@@ -859,59 +859,101 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
...
+                }
+            }
```
Comment (codemime) [inline src/overmap_ui.cpp]: I'd extract this into a separate function. `display` is already way too big.

File: src/overmap_ui.cpp [hunk 6]
```diff
@@ -859,59 +859,101 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
...
+
+            catacurses::window w_preview = catacurses::newwin( 5, max_note_length - 4, 2, 5 );
+            catacurses::window w_preview_title = catacurses::newwin( 2, max_note_length + 1, 0, 0 );
+            catacurses::window w_preview_map = catacurses::newwin( 5, 5, 2, 0 );
+
+            auto update_note_preview = [&]( std::string note ) {
```
Comment (codemime) [inline src/overmap_ui.cpp]: Accepting `note` by value.

File: src/overmap_ui.cpp [hunk 7]
```diff
@@ -859,59 +859,101 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
...
+                draw_border( w_preview_map, c_yellow );
+                for( int i = 1; i < 4; i++ ) {
+                    for( int j = 1; j < 4; j++ ) {
+                        const auto &ter = map_around.at( ( i - 1 ) * 3 + j - 1 );
+                        mvwputch( w_preview_map, i, j, ter.first, ter.second );
+                    }
+                }
+                mvwputch( w_preview_map, 2, 2, note_color, symbol );
+                wrefresh( w_preview_map );
+            };
```
Comment (codemime) [inline src/overmap_ui.cpp]: This one is a bit too big for being a lambda.

File: src/overmap_ui.cpp [hunk 8]
```diff
@@ -7,6 +7,8 @@
 #include "game.h"
 #include "input.h"
 #include "line.h"
+#include "map_iterator.h"
+#include "map.h"
```
Comment (codemime) [inline src/overmap_ui.cpp]: Why these headers? There's no `map` here.

File: src/overmap_ui.cpp [hunk 9]
```diff
@@ -82,6 +84,68 @@ std::tuple<char, nc_color, size_t> get_note_display_info( const std::string &not
     return result;
 }
 
+std::array<std::pair<nc_color, long>, 9> get_overmap_neighbors( const tripoint &current )
```
Comment (codemime) [inline src/overmap_ui.cpp]: It would be better to define a couple of local static constants: width and height of the neighborhood. This would eliminate the magic numbers and make the code that rely on them more robust.

File: src/overmap_ui.cpp [hunk 10]
```diff
@@ -82,6 +84,68 @@ std::tuple<char, nc_color, size_t> get_note_display_info( const std::string &not
...
+        long ter_sym = ' ';
+        const bool see = has_debug_vision || overmap_buffer.seen( dest.x, dest.y, dest.z );
+        if( see ) {
+            // Only load terrain if we can actually see it
+            oter_id cur_ter = overmap_buffer.ter( dest );
+            ter_color = cur_ter->get_color();
+            ter_sym = cur_ter->get_sym();
+        } else {
+            ter_color = c_dark_gray;
+            ter_sym   = '#';
```
Comment (codemime) [inline src/overmap_ui.cpp]: Excessive blanks.

File: src/overmap_ui.cpp [hunk 11]
```diff
@@ -82,6 +84,68 @@ std::tuple<char, nc_color, size_t> get_note_display_info( const std::string &not
...
+    mvwprintz( *w_preview_title, 0, 0, c_yellow, note_text );
+    mvwputch( *w_preview_title, 0, note_text.length(), c_white, LINE_XOXO );
+    for( size_t i = 0; i < note_text.length(); i++ ) {
+        mvwputch( *w_preview_title, 1, i, c_white, LINE_OXOX );
+    }
+    mvwputch( *w_preview_title, 1, note_text.length(), c_white, LINE_XOOX );
+    wrefresh( *w_preview_title );
+
+    draw_border( *w_preview_map, c_yellow );
+    for( int i = 1; i < 4; i++ ) {
```
Comment (codemime) [inline src/overmap_ui.cpp]: Starting from `0` would be more conventional and handy in case of named constants.

File: src/overmap_ui.cpp [hunk 12]
```diff
@@ -82,6 +84,68 @@ std::tuple<char, nc_color, size_t> get_note_display_info( const std::string &not
...
+        for( int j = 1; j < 4; j++ ) {
+            const auto &ter = map_around.at( ( i - 1 ) * 3 + j - 1 );
```
Comment (codemime) [inline src/overmap_ui.cpp]: `at` provides range checking. It looks like there's no need.

File: src/overmap_ui.cpp [hunk 13]
```diff
@@ -859,59 +923,47 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
 
             const std::string old_note = overmap_buffer.note( curs );
             std::string new_note = old_note, tmp_note;
+            std::array<std::pair<nc_color, long>, 9> map_around = get_overmap_neighbors( curs );
```
Comment (codemime) [inline src/overmap_ui.cpp]:
```suggestion
            const auto map_around = get_overmap_neighbors( curs );
```

File: src/overmap_ui.cpp [hunk 14]
```diff
@@ -859,59 +923,47 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
...
-                } else if( done ) {
+                } else if( input_popup.context().get_raw_input().get_first_input() == '\n' ) {
                     break;
+                } else {
+                    update_note_preview( new_note, map_around, preview_windows );
                 }
-
-            } while( true );
+            } while( input_popup.context().get_raw_input().get_first_input() != '\n' &&
+                     input_popup.context().get_raw_input().get_first_input() != KEY_ESCAPE );
```
Comment (codemime) [inline src/overmap_ui.cpp]: Why this change? There are `break`s for both keys inside the loop.

File: src/overmap_ui.cpp [hunk 15]
```diff
@@ -859,59 +923,47 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
...
 
+            do {
                 new_note = input_popup.query_string( false );
-                if( esc_pressed ) {
+                if( input_popup.context().get_raw_input().get_first_input() == KEY_ESCAPE ) {
                     new_note = old_note;
+                    esc_pressed = true;
                     break;
-                } else if( done ) {
+                } else if( input_popup.context().get_raw_input().get_first_input() == '\n' ) {
```
Comment (codemime) [inline src/overmap_ui.cpp]: I'd suggest extracting a temp. variable.

File: src/overmap_ui.cpp [hunk 16]
```diff
@@ -859,59 +923,47 @@ tripoint display( const tripoint &orig, const draw_data_t &data = draw_data_t()
...
-                                              get_tag_from_color( std::get<1>( om_symbol ) ),
-                                              std::get<0>( om_symbol ),
-                                              new_note.substr( std::get<2>( om_symbol ), std::string::npos ) );
-                } else {
-                    tmp_note.clear();
-                }
+            bool esc_pressed = false;
+            string_input_popup input_popup;
+            input_popup
+            .title( title )
```
Comment (codemime) [inline src/overmap_ui.cpp]: Some indentation wouldn't hurt here.

Comment (AMurkin) [reply src/overmap_ui.cpp]: Astyle want it unindented.

File: src/overmap_ui.cpp (hunk 2 above)
Comment (AMurkin) [reply src/overmap_ui.cpp]: It looks like `map::points_in_radius` works with different coordinates.
I'm getting one element when the cursor is far from player position.

Comment (AMurkin) [reply src/overmap_ui.cpp]: Fixed in dac51cc
